### PR TITLE
fix: align worker outputOffset with file size on PTY activation (closes #769)

### LINKED ISSUE #762: perf: client-side full-replay cost on session switch (xterm rebuild from cached bytes)
## Symptom

Switching between sessions produces a noticeable "full data replay" feel on every visit, even for the same session that was just visible. The browser tab visibly rebuilds the terminal contents from scratch each time the user clicks back to a session, which feels like the client is fetching the entire history from scratch.

## Investigation summary (2026-05-03 sprint start)

The initial assumption was that the **server fetch** is the cost. After tracing the Pull-model history flow (`packages/client/src/components/Terminal.tsx:864` → `packages/server/src/websocket/routes.ts:798`), it turned out the **server fetch is incremental** when the client cache is fresh: `requestHistory(sessionId, workerId, fromOffset = offsetRef.current)` sends a `request-history` with `fromOffset > 0`, and the server responds with only the delta. So the perceived "full fetch" is **not** a server-fetch cost.

The actual cost is on the **client side**:

1. `Terminal.tsx` is unmounted when the user switches away (its `useEffect` cleanup fires at `Terminal.tsx:784`, resetting `stateRef`, `offsetRef = 0`, `setCacheProcessed(false)`, aborting in-flight cache reads).
2. When the user switches back, the component remounts and re-runs `loadTerminalState(sessionId, workerId, signal)` (`Terminal.tsx:558`).
3. On cache hit, **the entire cached `data: string` (potentially MB-sized for long-running sessions) is replayed via `terminal.write(processOutputRef.current(cached.data), ...)`** (`Terminal.tsx:587`).
4. xterm.js rebuilds its internal scrollback buffer from those bytes — the visible "full re-render" cost.
5. Then the incremental `requestHistory(fromOffset)` is dispatched.

So even with a cache hit, every session re-visit pays the full xterm-write + buffer-rebuild cost for the cached content. The Pull-model fix (PR #655 / #648) reduced server fetches but did not eliminate the client-side replay.

## Trigger conditions (current behaviour)

| Condition | Replay cost |
|---|---|
| Session switch with cache hit, no truncation | Full client-side replay of cached bytes (server delta only) |
| Cache miss (first visit / 24h `MAX_AGE_MS` expired / cache cleared) | Server fetch of `WORKER_OUTPUT_INITIAL_HISTORY_LINES = 5000` lines + full client-side replay |
| Server-side truncation resync (`fromOffset > totalOffset` after 10 MB output-file truncation) | Server full-history resync + full client-side replay |
| Rapid back-and-forth between sessions | Full replay every switch (mount/unmount cycle re-runs each time) |

## Why the existing optimisations don't cover this

- **PR #655 (closes #648)** — eliminated cache wipe on server restart. Cache survives restart, but the unmount/remount cycle still re-replays cached bytes from scratch.
- **PR #311 (rapid tab-switching abort)** — cancels stale in-flight cache loads to avoid race conditions, but does not avoid the replay cost when the load completes.
- **PR #228 (idle-based save manager)** — saves cache periodically, but the read-and-replay path is unchanged.

The client-cache-as-bytes-string + xterm.write-on-mount design is the structural cause.

### A. Keep `<Terminal>` mounted on session switch (CSS visibility / `display: none` toggle)
- Pros: smallest scope, eliminates unmount/remount cycle entirely; xterm internal buffer survives between visits; no schema or persistence change.
- Cons: parent components (sessions list / route hierarchy) need to render all opened sessions concurrently — memory cost grows with number of opened sessions; lifecycle of inactive WebSocket connections needs explicit policy.
- Risk surface: WebSocket reconnect / disconnect logic interacts with inactive components.

### B. Persist xterm.js's serialised internal state (via `@xterm/addon-serialize`)
- Pros: compact representation; fast restore via `addon.serialize()` → snapshot string; cache content size shrinks; avoids replay-from-bytes.
- Cons: serialize addon's coverage of CJK / wide chars is the gap referenced in [#722](https://github.com/ms2sato/agent-console/issues/722); may require waiting on upstream fixes; replay path needs significant refactor.

### C. Web Worker / OPFS-backed terminal rendering
- Pros: terminal renderer lives in a Worker, browser can keep it alive across DOM unmounts; OPFS provides faster persistence than IndexedDB.
- Cons: large architectural change; xterm.js Worker support is partial; OPFS has narrower browser availability than IndexedDB.

### D. Reduce `terminal.write()` chunk size or yield via `requestIdleCallback`
- Pros: no architectural change; smooths perceived blocking time even though total cost is same.
- Cons: total cost unchanged; cosmetic improvement only; may make UX worse if input arrives during the chunked replay.

### Combination
B + A is also possible: serialise xterm state into the cache (smaller payload) AND keep the Terminal mounted (no replay cost at all). The two attack different parts of the cost budget.

## Acceptance Criteria (TBD per chosen approach)

To be decided after the owner reviews the candidate set above. Skeleton:

- [ ] Re-visiting a previously-visited session within the cache window completes within `<X> ms` (target: visually instant) — measured via DevTools performance trace.
- [ ] Memory growth across `N` opened sessions stays within `<Y> MB` (when Approach A is chosen, since unmounted components free memory).
- [ ] Existing cache-correctness invariants (`PR #655`'s "cache is owned by client; server's truncation-detection is the authority") remain intact.
- [ ] No regression on initial visit (cache miss) path — that path is fundamentally bounded by `WORKER_OUTPUT_INITIAL_HISTORY_LINES = 5000` and is unchanged by this Issue's scope.

## Related

- [#648](https://github.com/ms2sato/agent-console/issues/648) (closed via PR [#655](https://github.com/ms2sato/agent-console/pull/655)) — server-restart cache wipe, established the "cache is owned by client" contract
- [#397](https://github.com/ms2sato/agent-console/issues/397) (closed) — session-switch loading indicator UX
- [PR #311](https://github.com/ms2sato/agent-console/pull/311) — AbortController for rapid tab switching
- [#722](https://github.com/ms2sato/agent-console/issues/722) — vercel-labs/wterm adoption (gates serialize-addon adoption per Approach B)

## Why "腰を落ち着けて対応"

This Issue does not have a single obvious fix; the four candidate approaches have non-trivial trade-offs (memory vs persistence vs library availability), and the upstream serialize-addon CJK gap referenced in [#722](https://github.com/ms2sato/agent-console/issues/722) constrains Approach B's timeline. Implementation should be deferred to a future sprint where the design path is chosen explicitly, rather than rushed into a current sprint as a single-PR fix.

### LINKED ISSUE #769: design: align output offset semantics; fix disguised reload (#762)
## 背景

cache + offset 周辺は過去複数 PR でエンバグ再発の常連:

- [#648](https://github.com/ms2sato/agent-console/issues/648) — server restart cache wipe (PR #655 で fix)
- [PR #311](https://github.com/ms2sato/agent-console/pull/311) — rapid tab-switching abort
- [PR #228](https://github.com/ms2sato/agent-console/pull/228) — idle-based save manager
- [#762](https://github.com/ms2sato/agent-console/issues/762) — session 切替時 disguised 100% reload (本 Issue で fix)

root cause は **offset の semantic が server / client / event 種別で揺らいでいる** こと。 本 Issue で semantic と sequence を design 確定する。

## 現状の semantic 不整合

server 側で 1 worker に対して **2 種類の offset 計測が並走**:

| 由来 | 意味 | 設定箇所 |
|---|---|---|
| `worker.outputOffset` | **PTY activate 以降の累積 output bytes** (restoration で 0 reset) | `worker-manager.ts:552` 等 |
| `getCurrentOffset()` | **file 内の絶対 byte 位置** (永続) | `worker-output-file-manager.ts` |

ws event 種別で由来が異なる:

| ws event | offset 由来 | semantic |
|---|---|---|
| `output` | `worker.outputOffset` | cumulative (累積) |
| `history` | file size (`getCurrentOffset`) | file-absolute |
| `output-truncated` | `trimmedBuffer.length` | file-absolute |

新規 worker (file 空 + output flow) では両者一致するため問題が顕在化しないが、 **PTY restoration 後は乖離**。 client は両方を `offsetRef.current` 1 つの ref に代入 (Terminal.tsx L222 / L235) → race window で上書き → cache に壊れた offset 保存 → 次回 visit で server line-limit bypass で巨大 tail 再送 (= 「disguised 100% reload」)。

## Sequence (a) — normal happy path

```mermaid
sequenceDiagram
    autonumber
    participant U as User
    participant C as Client (Terminal.tsx)
    participant Cache as IndexedDB
    participant WS as WebSocket
    participant S as Server (worker-manager)
    participant F as Output file

    Note over S,F: Fresh worker creation
    S->>F: create file (size = 0)
    S->>S: worker.outputOffset = 0

    Note over U,C: Mount (1st visit)
    U->>C: navigate
    C->>Cache: loadTerminalState
    Cache-->>C: null (cache miss)
    C->>C: offsetRef = 0
    C->>WS: request-history(fromOffset=0)
    WS->>S: forward
    S->>F: read from 0 (line-limit 5000)
    F-->>S: data
    S->>WS: history(data, offset=file_size)
    WS->>C: forward
    C->>C: offsetRef = file_size
    C->>Cache: save(serialize, offset=file_size)

    Note over S,F: Live output
    S->>F: append chunk (file_size += chunk)
    S->>S: outputOffset += chunk
    Note right of S: outputOffset == file_size (fresh)
    S->>WS: output(chunk, offset=outputOffset)
    WS->>C: forward
    C->>C: offsetRef = offset
    C->>C: terminal.write(chunk)
    C->>Cache: markDirty (idle save)

    Note over U,C: Re-visit (cache hit)
    U->>C: navigate back
    C->>Cache: loadTerminalState
    Cache-->>C: { data: serialize, offset: file_size }
    C->>C: terminal.write(serialize) + offsetRef = file_size
    C->>WS: request-history(fromOffset=file_size)
    S->>F: read from file_size (incremental)
    F-->>S: empty / small delta
    S->>WS: history(delta, offset=new_size)
    Note right of C: ✅ ideal incremental sync
```

## Sequence (b) — restoration race window (#762 root cause)

```mermaid
sequenceDiagram
    autonumber
    participant U as User
    participant C as Client (Terminal.tsx)
    participant Cache as IndexedDB
    participant WS as WebSocket
    participant S as Server (worker-manager)
    participant F as Output file

    Note over S,F: Server restart / hibernation
    S->>S: PTY died, worker.pty = null
    Note right of F: file persists (size = N)

    Note over S: Activation (revived path)
    Note over S: restoreWorker / getAvailableWorker / pause-resume
    S->>S: spawn PTY
    S->>S: outputOffset = 0  ❌ should be N

    Note over U,C: Visit (cache hit from prev session)
    U->>C: navigate
    C->>Cache: loadTerminalState
    Cache-->>C: { data, offset: N }
    C->>C: offsetRef = N
    C->>WS: request-history(fromOffset=N)
    S->>F: read from N
    F-->>S: empty
    S->>WS: history(empty, offset=N)

    Note over S,F: PTY live output (large burst at activation)
    loop each chunk
        S->>F: append (file_size = N + delta)
        S->>S: outputOffset += chunk (small cumulative)
        S->>WS: output(chunk, offset=outputOffset)
        WS->>C: forward
        C->>C: offsetRef = outputOffset  ❌ overwrites N with small cumulative
    end
    C->>Cache: idle save (offset = small cumulative)  ❌ corrupted

    Note over U,C: Re-visit (cache hit, corrupted offset)
    U->>C: navigate away & back
    C->>Cache: loadTerminalState
    Cache-->>C: { data, offset: small_cumulative }
    C->>WS: request-history(fromOffset=small_cumulative)
    Note right of S: fromOffset > 0 → line-limit bypass
    S->>F: read from small_cumulative (file_size = N + delta)
    F-->>S: ~87% of file (huge tail)
    S->>WS: history(1.55MB delta, ...)
    Note right of C: ❌ disguised 100% reload
```

## 設計規約 (確定)

**全ての offset は file-absolute byte position (output ファイル先頭からのバイト位置)。**

| Phase | `worker.outputOffset` | `output.offset` | `history.offset` | `output-truncated.newOffset` |
|---|---|---|---|---|
| createWorker (fresh) | 0 (file size = 0) | file-absolute | file-absolute | — |
| live output | file size (advance per write) | file-absolute | file-absolute | — |
| PTY died (hibernation) | last value (`pty = null`) | — | file-absolute | — |
| **Activation (revived)** ⭐ | **`= getCurrentOffset()`** | file-absolute | file-absolute | — |
| restartWorker | 0 (file truncated to 0) | file-absolute | file-absolute | — |
| truncation (>10MB) | trimmed file size | file-absolute | file-absolute | trimmed file size |

⭐ = 現状 bug、 修正対象。

client `offsetRef` も同 semantic (file-absolute)。 race window 発生しても代入値が同 semantic なので壊れない。

### A. server-side seed (root cause fix)

`worker-manager` PTY activation path に `revived: boolean` flag を導入。 revived=true で `worker.outputOffset = await getCurrentOffset()`、 false で 0。

**Affected callers (revived = true)**:
- `worker-lifecycle-manager.ts:537-569` `restoreWorker()`
- `worker-lifecycle-manager.ts:217-279` `getAvailableWorker()` (`pty: null` activate path)
- `session-pause-resume-service.ts:233-266` paused session resume

**Affected callers (revived = false 維持)**:
- `worker-lifecycle-manager.ts:138-179` `createWorker()` (新規)
- `worker-lifecycle-manager.ts:389-420` `restartWorker()` (file truncate)

### B. client-side: 不要

codex review verdict — `offsetRef` は他 path (`saveCurrentTerminalState`, `registerSaveManager` getter) で消費するため `handleOutput` の `offsetRef = offset` 削除は別 bug を生む。

### C. design doc 追記

`docs/design/websocket-protocol.md` (or `session-worker-design.md`) に offset semantic 規約 + 上記 Mermaid 2 図を追記。

### D. Test

- `packages/server/src/services/__tests__/worker-lifecycle-manager.test.ts` — revived 各 path で activation 後 `worker.outputOffset === getCurrentOffset()` を assert
- `packages/server/src/services/__tests__/session-pause-resume-service.test.ts` — pause-resume path 同様
- 既存 client test (Terminal.test.tsx) は変更不要

## Acceptance Criteria

- [ ] PTY restoration / pause-resume / available-worker 全 3 path で activation 後 `worker.outputOffset === getCurrentOffset()` (file size) と一致 — server unit test
- [ ] dev 環境で session 開閉 3 回繰り返し、 2 回目以降の `request-history` data が **空 or 数 KB delta** に収まる (87% 巨大 tail が再送されない) — manual verification
- [ ] cache に保存される `offset` が常に file-absolute byte position と一致 — code review level
- [ ] design doc に offset semantic 規約 + Mermaid sequence 追記
- [ ] [#762](https://github.com/ms2sato/agent-console/issues/762) の symptom (session 切替時 disguised 100% reload) が再現しない — dogfood verification

## Follow-up (本 Issue scope 外、 別 Issue 候補)

1. `server-restarted` event の comment / 実装乖離 (`shared/src/types/session.ts:138` vs `client/src/lib/worker-websocket.ts:291-300` / `terminal-state-cache.ts:8-19`)
2. `output-truncated` poison path — server fix で同時解消されるが client 側にも guard 検討
3. 24h MAX_AGE_MS の適性 — 週末跨ぎで毎回 cache miss → line-limited 5000 行 reload

## Closes / Refs

- Closes #762
- Refs #648 #311 #228 (root cause 共通の cache + offset 周辺エンバグ)

### PR DESCRIPTION
**Pull request by ms2sato**
## Summary

- Thread a `revived: boolean` flag through `activateAgentWorkerPty` / `activateTerminalWorkerPty`. When true (the three revival paths — `restoreWorker`, `getAvailableWorker` PTY-null branch, `session-pause-resume` resume), seed `worker.outputOffset` from the persisted output file size before spawning the PTY. Fresh creation and `restartWorker` keep `revived: false` because their files are empty or have just been truncated.
- Document the offset semantic rule, lifecycle phase × offset table, and Mermaid sequences (happy path + the disguised-100%-reload regression mode) in `docs/design/websocket-protocol.md`. Add the missing `output-truncated` event row to the Server → Client message table.
- Cover all three revived paths with `worker-lifecycle-manager.test.ts` (real WorkerManager + memfs, asserts `outputOffset === getCurrentOffset()`) and pause-resume contract checks in `session-pause-resume-service.test.ts` (asserts `revived: true` reaches the worker-manager layer).

The server's `worker.outputOffset` was cumulative-from-PTY-activation, while the client persisted the file-absolute offset in IndexedDB. On revival the counter restarted from 0; the client's next `request-history` fed back the small cumulative number, hit the line-limit bypass, and the server replayed ~87% of the file as a fake "history" payload — the disguised 100% reload symptom of #762. Aligning both sides on the file-absolute semantic at activation time closes the race.

### TDD polarity check

1. New tests added → 5 fail against unmodified code (3 worker-lifecycle activation tests, 2 pause-resume contract tests).
2. Seed code added in `worker-manager.ts` → all 5 pass.
3. Reverted only the seed block (kept tests + call-site changes) → 3 worker-lifecycle activation tests fail again (pause-resume contract tests still pass because they only assert the call-site param, not the in-memory effect).
4. Restored seed block → all pass.

### CodeRabbit dispositions

- Round 1 / round 2 findings on adding try/catch + NaN/negative validation around `getCurrentOffset()`: skipped. `getCurrentOffset()` already catches its own errors and always returns a non-negative number; outer try/catch would be defensive coding for a scenario the API contract excludes (the design rules in `.claude/rules/design-principles.md` and `CLAUDE.md` "Avoid unnecessary complexity" warn against this).
- Doc finding (`output-truncated` event missing from ws protocol table): fixed — added the row in commit `9f881f6`.
- Test sync/async inconsistency: fixed — every `activate*Pty` test call now awaits, every enclosing `it(...)` is async (commit `8e2f644`).

Final CodeRabbit verdict: 0 findings.

## Test plan

- [x] `bun run typecheck` exits 0.
- [x] `bun run test` exits 0 — 4573 pass / 3 skip / 0 fail across all packages.
- [x] `bun run check:lang` clean (103 files scanned).
- [x] `coderabbit review --agent --base main` returns 0 findings.
- [ ] dev environment session-open / close 3 times: 2nd and later `request-history` payloads are empty or KB-scale delta (not the 87% MB tail).
- [ ] #762 symptom (disguised 100% reload on session switch) does not reproduce on dogfood.

Closes #769
Closes #762

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/server/src/services/__tests__/session-pause-resume-service.test.ts [hunk 1]
```diff
@@ -410,6 +410,83 @@ describe('SessionPauseResumeService', () => {
...
+          findById: mock(async () => persisted),
+          update: mock(async () => true),
+        },
+        workerManager: {
+          killWorker: mock(async () => {}),
+          restoreWorkersFromPersistence: mock(() => restoredWorkers),
+          activateAgentWorkerPty: activateAgentMock,
+          activateTerminalWorkerPty: mock(async () => {}),
+        } as unknown as SessionPauseResumeDeps['workerManager'],
+      });
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/services/__tests__/session-pause-resume-service.test.ts:447]: Addressed in commit 110e134:

Approach: narrowed `SessionPauseResumeDeps.workerManager` from `WorkerManager` (full class) to `Pick<WorkerManager, 'killWorker' | 'restoreWorkersFromPersistence' | 'activateAgentWorkerPty' | 'activateTerminalWorkerPty'>`, then replaced all 8 inline `as unknown as ...` casts with `satisfies`. With the dep type structurally narrowed to the four methods this service actually uses, the inline mock objects now structurally match without widening, so `satisfies` verifies shape compatibility while preserving the concrete `Mock<...>` types (which the new `revived: true` assertion tests rely on for typing `mock.calls[0][1]`).

Side benefit: the dep contract is now explicit about which `WorkerManager` methods this service consumes — adding a fifth call-site in production code will fail the test fixture's `satisfies` check, surfacing the contract change at the test boundary.

Local `coderabbit review --agent --base main` returns 0 findings on this commit.

**Comment by ms2sato** [comment]
@coderabbitai please review the latest changes — commit 110e134 addresses the previous CHANGES_REQUESTED finding by narrowing the dep type and switching to `satisfies`.
